Riot/Managers/URLPreviews/URLPreviewManager.swift

```diff
@@ -0,0 +1,148 @@
+//
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+
+@objcMembers
+/// A manager for URL preview data to handle fetching, caching and clean-up
+/// as well as remembering which previews have been closed by the user.
+class URLPreviewManager: NSObject {
+    /// The shared manager object.
+    static let shared = URLPreviewManager()
+    
+    /// A persistent store backed by Core Data to reduce network requests
+    private let store = URLPreviewStore()
+    
+    private override init() { }
+    
+    /// Generates preview data for a URL to be previewed as part of the supplied event,
+    /// first checking the cache, and if necessary making a request to the homeserver.
+    /// You should call `hasClosedPreview` first to ensure that a preview is required.
+    /// - Parameters:
+    ///   - url: The URL to generate the preview for.
+    ///   - event: The event that the preview is for.
+    ///   - session: The session to use to contact the homeserver.
+    ///   - success: The closure called when the operation complete. The generated preview data is passed in.
+    ///   - failure: The closure called when something goes wrong. The error that occured is passed in.
+    func preview(for url: URL,
+                 and event: MXEvent,
+                 with session: MXSession,
+                 success: @escaping (URLPreviewData) -> Void,
+                 failure: @escaping (Error?) -> Void) {
+        // Sanitize the URL before checking the store or performing lookup
+        let sanitizedURL = sanitize(url)
+        
+        // Check for a valid preview in the store, and use this if found
+        if let preview = store.preview(for: sanitizedURL, and: event) {
+            MXLog.debug("[URLPreviewManager] Using cached preview.")
+            success(preview)
+            return
+        }
+        
+        // Otherwise make a request to the homeserver to generate a preview
+        session.matrixRestClient.preview(for: sanitizedURL, success: { previewResponse in
+            MXLog.debug("[URLPreviewManager] Cached preview not found. Requesting from homeserver.")
+            
+            if let previewResponse = previewResponse {
```

Comment:
If there is no preview response neither `success` nor `failure` blocks are called. Maybe:
```swift
guard let previewResponse = previewResponse else {
  failure(error)
  return
}
```

Reply:
Nice catch, thanks 🙏